docs/_docs/20-fasterkv-basics.md

```diff
@@ -109,6 +109,74 @@ Apart from Key and Value, the IFunctions interface is defined on three additiona
 Callbacks for in-place updates receive the logical address of the record, which can be useful for applications such as indexing, and a reference to the `RecordInfo` header of the record, for use with the new locking calls.
 ReadCompletionCallback receives the `RecordInfo` of the record that was read.
 
+##### Return Values of IFunctions Data Operations
+IFunctions calls receive data through reference parameters to Keys and their associated Values. As well, they receive a parameter specific to the operation, describing additional information about the operation and its context:
+- **ReadInfo**: Contains the session Version, session type, and logical address
+- **UpsertInfo**, **RMWInfo**, **DeleteInfo**: Contains the session Version, session type, and logical address, as well as "output" variables for expiring records and canceling the operation:
+  - **DeleteRecord**: Always False coming in from FASTER; may be set True by the app if, for example, expiration has been detected. In this case, FASTER will set the Tombstone. It will assume deletion logic has already been done; that is, it will not call SingleDeleter or ConcurrentDeleter.
+  - **CancelOperation**: Always False coming in from FASTER; may be set True by the app if it detects a condition in the callback that makes the operation no longer desired.
+
+**Read** (`ReadInfo`):
+- `SingleReader`:
+  - ***Expiration***: The app sets readInfo.DeleteRecord true and returns false. FASTER will perform a logical delete of the record by forcing a CopyToTail of the current record with the Tombstone set, and Read will return (!status.Found && status.CreatedRecord).
+  - ***Cancellation***: The app sets readInfo.CancelOperation True and returns false. Read() returns (status.Canceled).
+  - ***Other false return***: Read() returns (status.NotFound).
+- `ConcurrentReader`
+  - ***Expiration***: The app sets readInfo.DeleteRecord true and returns false. FASTER will perform a logical delete of the record by in-place update to set the tombstone, and Read will return (status.NotFound | status.Record.InPlaceUpdated).
+  - ***Cancellation***: The app sets readInfo.CancelOperation True and returns false. Read() returns (status.Canceled).
+  - ***Other false return***: Read() returns (status.NotFound).
+
+**RMW** (`RMWInfo`)
+  - `InPlaceUpdater`
+    - ***Expiration***: The app sets rmwInfo.DeleteRecord true and returns false. FASTER will handle this as if the record was not found, and will enter the `NeedInitialUpdater` path, unless ***Cancellation*** was also done; in that case, FASTER will set recordInfo.Tombstone to true and will return (status.NotFound | status.InPlaceUpdated).
```

Comment:
If cancellation was not done, but DeleteRecord is set to true, IPU will still set the recordInfo.Tombstone bit before calling NIU, correct?